File: AnPham/Codeforces/1195D.py
MOD = 998244353

n = int(input())
a = input().split(' ')
a = list(map(lambda x: int(x), a))

a.sort()
la = list(map(lambda x: len(str(x)), a))
cnt = [0]*11
for val in la:
  cnt[val] += 1

def demSau(val, l):
  val = str(val)[-l:]
  res = ''
  for v in val:
    res += v*2
  return int(res)

def demTrc(val, l):
  val = int(str(val)[:-l])
  return val * pow(10, 2*l)

res = 0
for i in range(n):
  val = a[i]
  m = la[i]
  for l in range(11):
    res = (res + demSau(val, l) * cnt[l]) % MOD
  for l in range(1, m):
    res = (res + demTrc(val, l) * cnt[l] * 2) % MOD

print(res)

File: AnPham/Codeforces/1195D.cpp
#include <cstdio>
#include <algorithm>
#include <cstring>
#include <iostream>

using namespace std;

const int N = 100005;
const int inf = 1000000000;
const int MOD = 998244353;

typedef long long ll;

int n, a[N], cnt[15];
ll mu[50];

int getLength(int x) {
  int res = 0;
  for (; x > 0; x /= 10) res++;
  return res;
}

string toString(int val) {
  string s = "";
  while (val > 0) {
    s += (val % 10) + '0';
    val /= 10;
  }
  reverse(s.begin(), s.end());
  return s;
}

ll demSau(int a, int l) {
  string s = toString(a);
  s = s.substr(max(0, (int)s.length() - l), l);
  ll res = 0;
  for (int i = 0; i < s.length(); ++i) {
    res = (res*10 % MOD + s[i] - '0') % MOD;
    res = (res*10 % MOD + s[i] - '0') % MOD;
  }
  return res;
}

ll demTrc(int a, int l) {
  string s = toString(a);
  s = s.substr(0, s.length() - l);
  a = 0;
  for (int i = 0; i < s.length(); ++i)
    a = a * 10 + s[i] - '0';
  return (ll)a * mu[2*l] % MOD;
}

int main() {
  freopen("input.txt", "r", stdin);
  scanf("%d",&n);
  for (int i = 1; i <= n; ++i) {
    scanf("%d",&a[i]);
    int m = getLength(a[i]);
    cnt[m]++;
  }
  sort(a+1, a+n+1);
  mu[0] = 1;
  for (int i = 1; i <= 30; ++i)
    mu[i] = mu[i-1] * 10 % MOD;
  ll res = 0;
  for (int i = 1; i <= n; ++i) {
    int m = getLength(a[i]);
    for (int l = 1; l <= 11; ++l) {
      res = (res + demSau(a[i],l)*cnt[l] % MOD) % MOD;
    }
    for (int l = 1; l < m; ++l) {
      res = (res + demTrc(a[i],l)*cnt[l]%MOD * 2 % MOD) % MOD;
    }
  }
  cout << res << endl;
  return 0;
}